### DISCUSSION TITLE
fromLivewire method infinite loop

### DISCUSSION THREAD

**Discussion by suleymanozev** [spatie/laravel-data#168] [Q&A]
```php
    /**
     * @return array<string, mixed>
     */
    public function toLivewire(): array
    {
        return $this->all();
    }

    /**
     * @param  array<string, mixed>  $value
     *
     * @return static
     */
    public static function fromLivewire($value): static
    {
        return static::from($value);
    }
```

**Comment by rubenvanassche** [comment]
Could you please create a failing test case showing this error? This is not enough information to even start debugging the error.

**Comment by gehrisandro** [comment]
It's already a bit older but ran into the same issue today.

Make sure you are using the `withoutMagicalCreationFrom` method to prevent this infinite loop:

```php
public static function fromLivewire($value)
{
    return static::withoutMagicalCreationFrom($value);
}
```

**Reply by rubenvanassche** [reply]
Hi @gehrisandro, is this something you think we could fix in the package. haven't used data with LiveWire yet so I have no idea what's going on 😄

**Reply by gehrisandro** [reply]
Hi @rubenvanassche 

Good question! Before I found the solution mentioned above I was already preparing a PR to "fix" the package. But the solution I came up with wasn't that good. Basically I've added `fromLivewire` as an additional exception here: https://github.com/spatie/laravel-data/blob/main/src/Support/DataMethod.php#L25

In my project I've created a trait which I can apply to all DataObjects I'm going to use in Livewire:

```php
<?php

namespace App\Data\Traits;

trait WireableData
{
    public function toLivewire(): array
    {
        return $this->toArray();
    }

    public static function fromLivewire($value): static
    {
        return static::withoutMagicalCreationFrom($value);
    }
}
```

```php
<?php

namespace App\Data;

use App\Data\Traits\WireableData;
use Spatie\LaravelData\Data;

class SettingsData extends Data implements Wireable
{
    use WireableData;

    public function __construct(
        // ...
    ) {
    }
}
```

As you can see, it does not require that much effort to bring it to work with Livewire, but I guess most developers will stumble into the same problem. Would be nice to have a solution directly integrated into the package or at least a hint in the documentation.

Here is the Livewire documentation describing the interface: https://laravel-livewire.com/docs/2.x/properties#wireable-properties

**Comment by rubenvanassche** [comment]
I've added some extra Livewire support to the package + some extra docs on the matter with release 2.0.15!

**Reply by gehrisandro** [reply]
That's excellent 🎉 

Thank you very much!
